File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def can_place(i, j, mat, park):
    H, W = len(park), len(park[0])
    
    if(i+mat > H or j+mat > W):
        return False
    else:
        for p in range(i, i+mat):
            for q in range(j, j+mat):
                if(park[p][q] != '-1'):
                    return False
        return True
                
def solution(mats, park):
    H, W = len(park), len(park[0])
    mats.sort(reverse=True)
    
    for mat in mats:
        for i in range(H):
            for j in range(W):
                if(can_place(i, j, mat, park)):
                    return mat
    return -1
                
            
            

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool can_place(int i, int j, vector<vector<string>>& park,int mat) {
    int H = park.size();
    int W = park[0].size();
    if(i+mat > H || j+mat > W) return false;
    
    for(int p=i; p<i+mat; p++) {
        for(int q=j; q<j+mat; q++) {
            if(park[p][q] != "-1") return false;
        }
    }
    return true;
}

int solution(vector<int> mats, vector<vector<string>> park) {
    int H = park.size();
    int W = park[0].size();
    sort(mats.begin(), mats.end(), greater<int>());
    
    for(int mat : mats) {
        for(int i=0; i<H; i++) {
            for(int j=0; j<W; j++) {
                if(can_place(i, j, park, mat)) return mat;
            }
        }
    }
    
    return -1;
}